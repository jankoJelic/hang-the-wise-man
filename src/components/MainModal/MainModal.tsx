import Button from "components/Button";
import Subtitle from "components/textComponents/Subtitle";
import Title from "components/textComponents/Title";

type Props = {
  visible: boolean;
  title: string;
  onClickButton: () => void;
  description?: string;
  showTwoButtons?: boolean;
  confirmButtonTitle?: string;
  onClickCancel?: () => void;
};

const MainModal = ({
  visible,
  title,
  onClickButton,
  description,
  showTwoButtons = false,
  confirmButtonTitle = "OK",
  onClickCancel,
}: Props) => {
  return visible ? (
    <div
      className="flex fixed inset-0 left-0 w-screen h-screen
     bg-mainText bg-opacity-60 items-center justify-center shadow-md"
    >
      <div className="flex flex-col w-72 sm:w-96 p-4 h-54 sm:h-72 bg-white rounded-xl items-center relative">
        <Title text={title} />
        {description && <Subtitle text={description} />}
        <div className="absolute bottom-10 flex items-center">
          <Button
            onClick={onClickButton}
            styles="w-34 mx-2"
            title={confirmButtonTitle}
          />
          {showTwoButtons && (
            <Button
              onClick={onClickCancel}
              styles="w-34 bg-gray-600 mx-2 hover:bg-gray-400"
              title="Cancel"
            />
          )}
        </div>
      </div>
    </div>
  ) : (
    <></>
  );
};

export default MainModal;
