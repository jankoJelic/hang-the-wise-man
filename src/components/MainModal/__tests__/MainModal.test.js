import { fireEvent, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";

import MainModal from "../MainModal";

const mockClickConfirmButton = jest.fn();
const mockCancelClick = jest.fn();

const renderMainModal = () =>
  render(
    <MainModal onClickButton={mockClickConfirmButton} title="Modal" visible />
  );

const renderModalWithTwoButtons = () =>
  render(
    <MainModal
      onClickButton={mockClickConfirmButton}
      onClickCancel={mockCancelClick}
      confirmButtonTitle="Submit"
      title="Modal"
      visible
      showTwoButtons
    />
  );

it("renders correctly", () => {
  const tree = renderer.create(<MainModal />).toJSON();
  expect(tree).toMatchSnapshot();
});

describe("buttons", () => {
  it("shows one button by default", async () => {
    const wrapper = renderMainModal();
    expect(await wrapper.findAllByRole("button")).toHaveLength(1);
  });

  it("contains two buttons when 'showTwoButtons' prop is true", async () => {
    const wrapper = renderModalWithTwoButtons();
    expect(await wrapper.findAllByRole("button")).toHaveLength(2);
  });
});

describe("onClick handlers", () => {
  it("calls desired functions when buttons are being clicked", async () => {
    const wrapper = renderModalWithTwoButtons();

    const confirmButton = wrapper.getByText("Submit");
    const cancelButton = wrapper.getByText("Cancel");

    fireEvent.click(confirmButton);
    fireEvent.click(cancelButton);
    fireEvent.click(cancelButton);

    expect(mockClickConfirmButton).toHaveBeenCalledTimes(1);
    expect(mockCancelClick).toHaveBeenCalledTimes(2);
  });
});
