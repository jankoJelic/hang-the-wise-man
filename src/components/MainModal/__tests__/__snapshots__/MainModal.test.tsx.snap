// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="flex fixed inset-0 left-0 w-screen h-screen
   bg-mainText bg-opacity-60 items-center justify-center shadow-md undefined"
>
  <div
    className="flex flex-col w-72 sm:w-96 p-4 h-54 sm:h-72 bg-white rounded-xl items-center relative"
  >
    <h1
      className="font-bold text-2xl sm:text-4xl text-mainText text-center mb-8"
    >
      Modal
    </h1>
    <div
      className="absolute bottom-10 flex items-center"
    >
      <button
        className="bg-mainAction text-white w-full max-w-md h-10 rounded-xl hover:bg-mainLight mb-2 font-semibold w-34 mx-2"
        disabled={false}
        onClick={[MockFunction]}
      >
        OK
      </button>
    </div>
  </div>
</div>
`;
